service/frontend/templates/ratelimited.tmpl

```diff
@@ -0,0 +1,119 @@
+
+{{$domainIDAPIs := list "RecordActivityTaskHeartbeat" "RespondActivityTaskCanceled" "RespondActivityTaskCompleted" "RespondActivityTaskFailed" "RespondDecisionTaskCompleted" "RespondDecisionTaskFailed" "RespondQueryTaskCompleted"}}
+{{$queryTaskTokenAPIs := list "RespondQueryTaskCompleted"}}
+{{$nonBlockingAPIs := list "RecordActivityTaskHeartbeat" "RecordActivityTaskHeartbeatByID" "RespondActivityTaskCompleted" "RespondActivityTaskCompletedByID" "RespondActivityTaskFailed" "RespondActivityTaskFailedByID" "RespondActivityTaskCanceled" "RespondActivityTaskCanceledByID" "RespondDecisionTaskCompleted" "RespondDecisionTaskFailed" "RespondQueryTaskCompleted" "ResetStickyTaskList"}}
+
+{{$interfaceName := .Interface.Name}}
+{{$handlerName := (index .Vars "handler")}}
+{{ $decorator := (printf "%s%s" (down $handlerName) $interfaceName) }}
+{{ $Decorator := (printf "%s%s" $handlerName $interfaceName) }}
+
+// {{$decorator}} implements {{.Interface.Type}} interface instrumented with rate limiter.
+type {{$decorator}} struct {
+    wrapped     {{.Interface.Type}}
+    tokenSerializer common.TaskTokenSerializer
+    domainCache cache.DomainCache
+    userRateLimiter quotas.Policy
+    workerRateLimiter quotas.Policy
+    visibilityRateLimiter quotas.Policy
+}
+
+// New{{$Decorator}} creates a new instance of {{$interfaceName}} with ratelimiter.
+func New{{$Decorator}}(
+    wrapped {{.Interface.Type}},
+    domainCache cache.DomainCache,
+    userRateLimiter quotas.Policy,
+    workerRateLimiter quotas.Policy,
+    visibilityRateLimiter quotas.Policy,
+) {{.Interface.Type}} {
+    return &{{$decorator}}{
+        wrapped: wrapped,
+        tokenSerializer: common.NewJSONTaskTokenSerializer(),
+        domainCache: domainCache,
+        userRateLimiter: userRateLimiter,
+        workerRateLimiter: workerRateLimiter,
+        visibilityRateLimiter: visibilityRateLimiter,
+    }
+}
+
+{{range $method := .Interface.Methods}}
+func (h *{{$decorator}}) {{$method.Declaration}} {
+    {{- if hasKey $ratelimitTypeMap $method.Name}}
+        {{- $ratelimitType := get $ratelimitTypeMap $method.Name}}
+        {{- $domain := printf "%s.GetDomain()" (index $method.Params 1).Name}}
+        {{- $rateLimitHelper := "allowDomain"}}
+        {{- if has $method.Name $domainIDAPIs}}
+            {{- $domain = "domainName"}}
+            {{- $rateLimitHelper := "allowDomainID"}}
+            {{- if has $method.Name $queryTaskTokenAPIs}}
+                token, err := h.tokenSerializer.DeserializeQueryTaskToken({{(index $method.Params 1).Name}}.TaskToken)
+            {{- else}}
+                token, err := h.tokenSerializer.Deserialize({{(index $method.Params 1).Name}}.TaskToken)
+            {{- end}}
+            if err != nil {
+                return
+            }
+            domainName, err := h.domainCache.GetDomainName(token.DomainID)
+            if err != nil {
+                return
+            }
```

Comment:
```
            {{- $rateLimitHelper := "allowDomainID"}}
```
with `:=`, this looks unused in this block.  or does it still modify the lines below that have this?
```
            h.{{$rateLimitHelper}}({{$ratelimitType}}, {{$domain}})
```

(I'm not 100% sure what the scoping behavior is with templates, but I assume it's the same as go...?  hopefully?)

Reply:
it's unused